proposals/0000-swift-turbo-modules.md

```diff
@@ -0,0 +1,185 @@
+---
+title: Swift Turbo Modules
+author:
+- Oskar Kwasniewski 
+- Riccardo Cipolleschi
+date: 12-08-2024 
+---
+
+# RFC0000: Swift Turbo Modules
+
+## Summary
+
+This RFC aims to allow developers to write Turbo Modules using Swift. This will allow the usage of more modern language making maintenance of native modules easier and more accessible. 
+
+## Motivation
+
+The primary motivations for introducing Swift Turbo Modules are:
+- Enhance developer experience for iOS developers working with React Native
+- Allow the use of more modern language
+- Lower the entry barrier to write a native turbo module
+
+
+## Detailed design
+
+One of the reasons why we can't adopt Swift in TurboModules is the contamination of C++ types ending up in user-space.
```

Comment:
While i can understand not wanting to expose internal C++ types to user land, reality is that RN is based on C++, and if something should be exposed and used by library maintainers, then having a clear API that defines public vs internal is all that is required, the language is inconsequential. I think as a library developer, regardless of what we do, we will eventually have to debug or deal with those internals, and this should be fine. Though maybe its an unpopular opinion

Reply:
I completely agree with you. But we would like to provide a cleaner and simple experience to library maintainers. Not every library needs to know how react native works under the hoods, no?

We also have an effort internally to try and define the public boundary better, with a clear distinction between what's public and what's private. The project is almost 10 years old in OSS, so work is not trivial and requires a bit of time, but we hope to get there soon!